From: Amit Shah <amit.shah@redhat.com>
Date: Sat, 7 Aug 2010 07:16:52 -0400
Subject: [virtio] initialize vq->data entries to NULL
Message-id: <9eea5d30bd0906a2fba5f0edcd0713d9a8330f39.1281129816.git.amit.shah@redhat.com>
Patchwork-id: 27467
O-Subject: [RHEL5.6 PATCH 7/9] virtio: Initialize vq->data entries to NULL
Bugzilla: 620037
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

vq operations depend on vq->data[i] being NULL to figure out if the vq
entry is in use (since the previous patch).

We have to initialize them to NULL to ensure we don't work with junk
data and trigger false BUG_ONs.

Signed-off-by: Amit Shah <amit.shah@redhat.com>
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
Cc: Shirley Ma <xma@us.ibm.com>
(cherry picked from commit 3b8706240ee6084ccb46e53cd3a554356b7eeec8)

Signed-off-by: Amit Shah <amit.shah@redhat.com>

diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index 2f97f58..8d43699 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -356,8 +356,11 @@ struct virtqueue *vring_new_virtqueue(unsigned int num,
 	/* Put everything in free lists. */
 	vq->num_free = num;
 	vq->free_head = 0;
-	for (i = 0; i < num-1; i++)
+	for (i = 0; i < num-1; i++) {
 		vq->vring.desc[i].next = i+1;
+		vq->data[i] = NULL;
+	}
+	vq->data[i] = NULL;
 
 	return &vq->vq;
 }
